### PR TITLE
root 6.28.04

### PR DESCRIPTION
**Pull request by henryiii**
- minuit2 6.26.10
- root 6.26.10

(Both need to be bumped at the same time)

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

I got reports of ACLiC being broken on a "recent" macOS, and tried it on macOS 13, and it seems this is broken (actually more than just ACLiC might be, I get messages about missing libraries). I added a test recommended by the ROOT team, as well as a check for ACLiC. I can't test this locally because it fails building for me on MacOS 13. Hopefully this is a good step forward for non-13 macOS.

### CODE REVIEW

File: Formula/root.rb [hunk 1]
```diff
@@ -149,6 +149,10 @@ def caveats
     assert_equal "\nProcessing test.C...\nHello, world!\n",
                  shell_output("root -l -b -n -q test.C")
 
+    # Test ACLiC
+    assert_equal "\nProcessing test.C...\nHello, world!\n",
```
Comment (henryiii) [inline Formula/root.rb:153]:
```suggestion
    assert_equal "\nProcessing test.C+...\nHello, world!\n",
```

**Comment by henryiii** [comment]
FYI, the recommendation is to set all built-ins to off (except LLVM & Clang). This is what the conda-forge package uses:

```
      -Dbuiltin_afterimage=OFF
      -Dbuiltin_cfitsio=OFF
      -Dbuiltin_davix=OFF
      -Dbuiltin_fftw3=OFF
      -Dbuiltin_freetype=OFF
      -Dbuiltin_ftgl=OFF
      -Dbuiltin_gl2ps=OFF
      -Dbuiltin_glew=OFF
      -Dbuiltin_gsl=OFF
      -Dbuiltin_lz4=OFF
      -Dbuiltin_lzma=OFF
      -Dbuiltin_nlohmannjson=OFF
      -Dbuiltin_openssl=OFF
      
      -Dbuiltin_openui5=ON
      -Dbuiltin_pcre=OFF
      -Dbuiltin_tbb=OFF
      -Dbuiltin_unuran=OFF
      -Dbuiltin_vc=OFF
      -Dbuiltin_vdt=OFF
      -Dbuiltin_xrootd=OFF
      -Dbuiltin_xxhash=OFF
      -Dbuiltin_zlib=OFF
      -Dbuiltin_zstd=OFF

      -Dbuiltin_llvm=ON
      -Dbuiltin_clang=ON
```

I can't test locally yet due to the macOS 13 issue I'm having, so not adding those in CI just yet.

**Comment by Bo98** [comment]
> Looks like libncurses is gone in macOS 13?

Should be still there. The headers certainly are still in the SDK.

**Comment by henryiii** [comment]
Trying a local build, but the updated formula doesn't auto-download things that are missing, and fixed the issue with it not finding xrootd. The reported list of features at the end is the same (minus some "builtin" reports).

> Should be still there. The headers certainly are still in the SDK.

I don't see it in `/usr/lib`, though?

**Comment by Bo98** [comment]
Ah I see your problem: https://github.com/root-project/root/blob/25e425ff1186e9c9287d26467b66dce3b4cbf108/core/metacling/src/TCling.cxx#L3255-L3278

These paths changed in Ventura to `/System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_arm64e` or something.

If possible, I recommend not reading those internal files. The official way to do it is the `_dyld_shared_cache_contains_path` function.

**Comment by henryiii** [comment]
Someone's working on fixing ROOT on macOS 13. I passed on the info. Thanks!!!

**Comment by chenrui333** [comment]
Let us know when we can do the rerun, mark the PR as `in progress`

**Comment by henryiii** [comment]
ROOT just fixed the issues with macOS 13 & released 6.26.10 a day or two ago. I'll update to that, I can also test locally now that Xcode is out.

**Comment by henryiii** [comment]
I've updated, but local builds are failing with:

```
FAILED: lib/libNetxNG.so
: && /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -std=c++17 -Wc++11-narrowing -Wsign-compare -Wsometimes-uninitialized -Wconditional-uninitialized -Wheader-guard -Warray-bounds -Wcomment -Wtautological-compare -Wstrncat-size -Wloop-analysis -Wbool-conversion -m64 -pipe -W -Wall -Woverloaded-virtual -fsigned-char -fno-common -Qunused-arguments -pthread -stdlib=libc++ -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.sdk -dynamiclib -Wl,-headerpad_max_install_names -m64 -single_module -Wl,-dead_strip_dylibs -Wl,-rpath,/usr/local/Cellar/root/6.26.10/lib/root -o lib/libNetxNG.so -install_name @rpath/libNetxNG.so net/netxng/CMakeFiles/G__NetxNG.dir/G__NetxNG.cxx.o net/netxng/CMakeFiles/NetxNG.dir/src/TNetXNGFile.cxx.o net/netxng/CMakeFiles/NetxNG.dir/src/TNetXNGFileStager.cxx.o net/netxng/CMakeFiles/NetxNG.dir/src/TNetXNGSystem.cxx.o net/netxng/CMakeFiles/NetxNG.dir/src/RRawFileNetXNG.cxx.o  -Wl,-rpath,/tmp/root-20221118-97723-17dfg1n/root-6.26.10/builddir/lib  lib/libNet.so  lib/libRIO.so  lib/libThread.so  /usr/local/lib/libXrdUtils.dylib  /usr/local/lib/libXrdClient.dylib  /usr/local/lib/libXrdCl.dylib  lib/libCore.so && :
Undefined symbols for architecture x86_64:
  "XrdCl::StatInfo::~StatInfo()", referenced from:
      TNetXNGFile::GetSize() const in TNetXNGFile.cxx.o
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
      TNetXNGSystem::Unlink(char const*) in TNetXNGSystem.cxx.o
      ROOT::Internal::RRawFileNetXNG::GetSizeImpl() in RRawFileNetXNG.cxx.o
  "XrdCl::StatInfo::GetModTime() const", referenced from:
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
  "XrdCl::StatInfo::GetId() const", referenced from:
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
  "XrdCl::StatInfo::GetSize() const", referenced from:
      TNetXNGFile::GetSize() const in TNetXNGFile.cxx.o
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
      ROOT::Internal::RRawFileNetXNG::GetSizeImpl() in RRawFileNetXNG.cxx.o
  "XrdCl::StatInfo::GetFlags() const", referenced from:
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
  "XrdCl::StatInfo::TestFlags(unsigned int) const", referenced from:
      TNetXNGSystem::GetPathInfo(char const*, FileStat_t&) in TNetXNGSystem.cxx.o
      TNetXNGSystem::Unlink(char const*) in TNetXNGSystem.cxx.o
ld: symbol(s) not found for architecture x86_64
```

Checking with the ROOT team.

**Comment by henryiii** [comment]
Could you trigger a build just to see if this also shows up (and where)? The ROOT team hasn't seen this failure before. It might be due to my xrootd build being from macOS 12 and the new build being on 13.

**Comment by chenrui333** [comment]
> Could you trigger a build just to see if this also shows up (and where)? The ROOT team hasn't seen this failure before. It might be due to my xrootd build being from macOS 12 and the new build being on 13.

doing that now.

**Comment by chenrui333** [comment]
some audit failures

```
  root:
    * Libraries were compiled with a flat namespace.
      This can cause linker errors due to name collisions, and
      is often due to a bug in detecting the macOS version.
        /opt/homebrew/Cellar/root/6.26.10/lib/root/libCling.so
        /opt/homebrew/Cellar/root/6.26.10/lib/root/libJupyROOT3_10.so
        /opt/homebrew/Cellar/root/6.26.10/lib/root/libROOTPythonizations3_10.so
        /opt/homebrew/Cellar/root/6.26.10/lib/root/libROOTTPython.so
        /opt/homebrew/Cellar/root/6.26.10/lib/root/libcppyy3_10.so
```

**Comment by chenrui333** [comment]
Looks like just some audit failures need to be addressed @henryiii

**Comment by Bo98** [comment]
Yes that's correct. Though note that `-undefined dynamic_lookup` is not the same thing as `-flat_namespace` that the audit is complaining about. Flat namespaces have effectively been deprecated (albeit not officially) since the move from PowerPC to Intel.

You will see a warning about `-undefined dynamic_lookup` in the latest Xcode. Ignore that for now as there's no proper replacement yet - I've talked to Apple directly about the warning and have been told to wait for an upcoming Xcode release.

Out of the ~6400 formulae we have, only about 20 build with `-flat_namespace` and some of them are from really old build setups that haven't been changed since the early 2000s.

I'm guessing the relevant changes here were https://github.com/root-project/root/commit/a05d4beded1481a86a65e450826352eb528b03e7 and https://github.com/root-project/root/commit/288dc3cf878a116e1a5657af29a1cbc38292420f.

**Comment by chenrui333** [comment]
okay, marking this `ready to merge` 👍

**Comment by Bo98** [comment]
It will need a PR to consider audit exception if flat namespace is the direction ROOT wants to take.

By "that's correct" I was replying to Henry's comment.

**Comment by henryiii** [comment]
I'll try reverting those two commits later. ROOT team says they don't care how it's done, but they require no warnings. `-undefined dynamic_lookup` produces a warning and `-flat_namespace` did not.

**Comment by Bo98** [comment]
Could be a case of waiting until Xcode 14.2 then.

If you want a temporary fix to silence warnings you could try `-Wl,-no_fixup_chains`, but I say that with the caveat that it is temporary, undocumented and should be removed when there's a proper fix in the future.

**Comment by cho-m** [comment]
There is an upstream fix for `-flat_namespace` issue which has been backported to 6.28 branch. Not sure how long until 6.28/00 release.

**Comment by branchvincent** [comment]
6.28/00 was just [released](https://root.cern.ch/releases/release-62800/)

**Comment by henryiii** [comment]
Thanks for getting to that, yes, I was waiting for 6.28.

**Comment by henryiii** [comment]
The standalone build of minuit2 is broken. As long as these are allowed to go out of sync, I think we can move forward on the ROOT build and leave the minuit2 one until something can be worked out.

For ROOT, external Davix seems to be broken (or requires an update? Haven't checked). But we can go to internal Davix for now.

**Comment by SMillerDev** [comment]
```
  /home/linuxbrew/.linuxbrew/Cellar/minuit2/6.28.02/include/Minuit2/Minuit2/MnMatrix.h:19:10: fatal error: Minuit2/MnMatrixfwd.h: No such file or directory

```

Interesting enough, different error on macOS.

**Comment by cho-m** [comment]
Added some patches, but we could just drop support for `minuit2` since no one seems to be using it and looks like upstream support for standalone is low priority given that it wasn't caught by some basic CI checks.

**Comment by SMillerDev** [comment]
Sounds good, are you up for that?

**Comment by cho-m** [comment]
> Is it possible to update root but not minuit2 for now? This might get fixed in a few months. Dropping it is fine by me too.

You can try removing
https://github.com/Homebrew/homebrew-core/blob/970acdcad3be4262451e9a5180a385dd2158eda3/synced_versions_formulae.json#L41

Not sure if you will hit some other audit check.

Also will probably need to combine #134838 since we have merged in `openssl@3` migration so users may hit some issues with multiple OpenSSL versions in use. Will also need to remove
https://github.com/Homebrew/homebrew-core/blob/970acdcad3be4262451e9a5180a385dd2158eda3/audit_exceptions/versioned_dependencies_conflicts_allowlist.json#L12

**Comment by cho-m** [comment]
Building bundled LLVM is hitting various issues. I'm guessing Linux failure is due to `llvm@15` headers getting mixed. Not sure on macOS failure but could be some SDK mixing.

Would need someone to look into what is causing this.

Alternatively, could try using Homebrew's `llvm@13` instead but will need https://github.com/root-project/root/pull/12153. This will lead to multiple LLVM versions on Linux but that is probably okay given we don't even have Linux bottle right now.

**Comment by cho-m** [comment]
Sadly Root is mixing some code from different LLVM versions so can't use `llvm@13`.

I think macOS failure is `-I/opt/homebrew/opt/util-linux/include` added by superenv mixing `uuid.h` headers.

**Review by p-linnane — APPROVED** [review]

**Pull request merged**